src/main/java/org/xbill/DNS/lookup/LookupSession.java

```diff
@@ -0,0 +1,314 @@
+ * LookupSession provides facilities to make DNS Queries. A LookupSession is intended to be long
+ * lived, and it's behaviour can be configured using the properties of the LookupSessionBuilder
+ * instance returned by the builder() method.
+ */
+@Builder
+public class LookupSession {
+  // remember to update javadoc if this is changed for the following two properties.
+  public static final int DEFAULT_MAX_ITERATIONS = 16;
+  public static final int DEFAULT_NDOTS = 1;
+  /** The {@link Resolver} to use to look up records. */
+  @NonNull private final Resolver resolver;
+  /**
+   * Sets the maximum number of CNAME or DNAME redirects allowed before lookups with fail with
+   * {@link RedirectOverflowException}. Defaults to {@code 16}.
+   */
+  @Builder.Default private final int maxRedirects = DEFAULT_MAX_ITERATIONS;
+  /**
+   * Sets the threshold for the number of dots which much appear in a name before it is considered
+   * absolute. The default is {@value #DEFAULT_NDOTS}, meaning meaning that if there are any dots in
+   * a name, the name will be tried first as an absolute name.
+   */
+  @Builder.Default private final int ndots = DEFAULT_NDOTS;
+
+  /** Configures the search path used to look up relative names with less than ndots dots. */
+  @Singular("searchPath")
+  private final List<Name> searchPath;
+
+  /**
+   * If set to {@code true}, cached results with multiple records will be returned with the starting
+   * point shifted one step per request.
+   */
+  @Builder.Default private final boolean cycleResults = false;
+
+  /**
+   * Configures the Cache instances to be used for lookups for the different {@link DClass} values.
+   */
+  @Singular("cache")
+  private final Map<Integer, Cache> caches;
+
+  /**
+   * A builder for {@link LookupSession} instances where functionality is mostly generated as
+   * described in the <a href="https://projectlombok.org/features/Builder">Lombok Builder</a>
+   * documentation. An instance of this class is obtained by calling {@link LookupSession#builder()}
+   * and configured using the methods with names corresponding to the different properties. Once
+   * fully configured, a {@link LookupSession} instance is obtained by calling {@link
+   * LookupSessionBuilder#build()} on the builder instance.
+   */
+  public static class LookupSessionBuilder {
+    void preBuild() {
+      // note that this transform is idempotent, as concatenating an already absolute Name with root
+      // is a noop.
+      if (searchPath != null) {
+        this.searchPath =
+            searchPath.stream()
+                .map(
+                    name -> {
+                      try {
+                        return Name.concatenate(name, Name.root);
+                      } catch (NameTooLongException e) {
+                        throw new IllegalArgumentException("Search path name too long");
+                      }
+                    })
+                .collect(Collectors.toCollection(ArrayList::new));
```

Comment:
The comment probably got lost, but what about `searchPath.stream()...().forEach(this.searchPath::add);` instead of ArrayList::new? It would avoid the creation of a never used list.

Reply:
Well, calling add() on a list from within a stream over that list causes a `ConcurerntModificationException`. While it is good to not create throwaway values left and right, I think that constructing a new list is okay in this case. Let's call it the least worst option.

Reply:
Ah right, I didn't pay attention that this is the same list.